### PR TITLE
keypaths: handle mainnet and testnet keypaths and coins correctly

### PR DESCRIPTION
**Pull request by Tomasvrba**

### CODE REVIEW

File: src/bitbox02.js [hunk 1]
```diff
@@ -140,11 +142,13 @@ export class BitBox02API {
         });
     }
 
-    // @return the eth xpub at m/44'/60'/0'/0
-    async ethGetRootPubKey() {
+    // @return the eth xpub for a given coin and derivation path
+    async ethGetRootPubKey(path) {
```
Comment (benma) [inline src/bitbox02.js]: please rename to keypath, less ambiguios

File: src/eth-utils.js [hunk 2]
```diff
@@ -1,5 +1,61 @@
...
+        default:
+            throw new Error('Unsupported network');
+    }
+}
+
+/**
+ * @param pathString keypath in string format e.g. m/44'/1'/0'/0
+ * @returns keypath as array e.g. [2147483692, 2147483649, 2147483648, 0]
+ */
+export const getPathFromString = pathString => {
```
Comment (benma) [inline src/eth-utils.js]: same, `getKeypathFromString`, `keypathString`.

File: src/eth-utils.js [hunk 3]
```diff
@@ -1,5 +1,61 @@
...
+    const levels = pathString.toLowerCase().split('/');
+    if (levels[0] !== 'm') throw new Error('Invalid keypath');
+    return levels.flatMap(level => {
+        if (level === 'm' || level === '') {
+            return [];
+        }
+        if (level.substring(level.length - 1) === "'") {
+            level = +level.substring(0, level.length - 1) + HARDENED;
```
Comment (benma) [inline src/eth-utils.js]: what if the number before `'` is already >= HARDENED?

File: src/eth-utils.js [hunk 4]
```diff
@@ -1,5 +1,61 @@
...
+        }
+        level = parseInt(level);
+        if (isNaN(level) || level < 0) throw new Error('Invalid keypath');
```
Comment (benma) [inline src/eth-utils.js]: there is also an upper limit afaik

File: src/eth-utils.js [hunk 5]
```diff
@@ -1,5 +1,61 @@
...
+        return level;
+    })
+}
+
+/**
+ * @param pathArray keypath as an array of ints e.g. [44, 1, 0, 0]; or hardened [2147483692, 2147483649, 2147483648, 0]
```
Comment (benma) [inline src/eth-utils.js]: remove unhardened part, haven't seen that used anywhere

File: src/eth-utils.js (hunk 3 above)
Comment (benma) [reply src/eth-utils.js]: what about an input like `m/-1'`? that should not work but maybe this is turned into an unhardened high index.

File: src/eth-utils.js [hunk 6]
```diff
@@ -1,5 +1,66 @@
...
+            return [];
+        }
+        if (level.substring(level.length - 1) === "'") {
+            hardened = true
+        }
+        level = parseInt(level);
+        if (isNaN(level) ||
+            level < 0 ||
+            (!hardened && level >= HARDENED) ||
+            (hardened && level >= HARDENED * 2)
```
Comment (benma) [inline src/eth-utils.js]: this check is still wrong i think; in `m/x'`, x still needs to be `< HARDENED`, as a couple of lines below you add `HARDENED` to it. so in both hardened and unhardened the check is the same here.

File: src/eth-utils.js [hunk 7]
```diff
@@ -1,5 +1,61 @@
...
+}
+
+/**
+ * @param keypathString keypath in string format e.g. m/44'/1'/0'/0
+ * @returns keypath as array e.g. [2147483692, 2147483649, 2147483648, 0]
+ */
+export const getKeypathFromString = keypathString => {
+    let levels = keypathString.toLowerCase().split('/');
+    if (levels[0] !== 'm') throw new Error('Invalid keypath');
+    levels = levels.slice(1, levels.length);
```
Comment (benma) [inline src/eth-utils.js]: nit: same as just `levels.splice(1)`, which deletes the first element without making a copy of the rest.

Comment (Tomasvrba) [reply src/eth-utils.js]: Good point.

**Pull request merged**